---
title: "class18_part2"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Protein sequences from healthy and tumor tissue

The following sequences resulted from such an NGS analysis of patient healthy and tumor tissue. 

Your task is to identify tumor specific mutations that could potentially be used for vaccine development.

> Q1. Identify sequence regions that contain all 9-mer peptides that are only found in the tumor. 

Hint: You will need to first identify the sites of mutation in the above sequences and then extract the surrounding subsequence region. This subsequence should encompass all possible 9-mers in the tumor derived sequence. In other words extract the subsequence from 8 residues before and 8 residues after all point mutations in the tumor sequence.

```{r}
library(bio3d)
seqs <- read.fasta("lecture18_sequences.fa")

# We can optionally align these sequences, but in this case the provided sequences are already aligned.
# seqs <- seqaln(seqs)

## will need to use MUSCLE to align
## download MUSCLE and call it in R terminal
```

Next we calculate identity per equivalent (i.e. aligned) position and then use this information to find non identical sites that do not contain gaps (i.e. indels).

```{r}
## Calculate positional identity scores
ide <- conserv(seqs$ali, method="identity")
mutant.sites <- which(ide < 1) 

## Exclude gap positions from analysis
gaps <- gap.inspect(seqs)
mutant.sites <- mutant.sites[mutant.sites %in% gaps$f.inds]

mutant.sites
```

We can use these indices in mutant.sites to extract subsequences as required for the hands-on session. First however we come up with suitable names for these subsequences based on the mutation. This will help us later to make sense and keep track of our results.

```{r}
## Make a "names" label for our output sequences (one per mutant)
mutant.names <- paste0(seqs$ali["P53_wt",mutant.sites],
                       mutant.sites,
                       seqs$ali["P53_mutant",mutant.sites])

mutant.names
```

Now lets extract all 9-mer mutant encompassing sequences for each mutant site. This is equivalent to finding the sequence region eight residues before and eight residues after our mutation sites and outputting this subsequence to a new FASTA file.

```{r}
## Sequence positions surounding each mutant site
start.position <- mutant.sites - 8
end.position <-  mutant.sites + 8

# Blank matrix to store sub-sequences
store.seqs <- matrix("-", nrow=length(mutant.sites), ncol=17)
rownames(store.seqs) <- mutant.names

## Extract each sub-sequence
for(i in 1:length(mutant.sites)) {
  store.seqs[i,] <- seqs$ali["P53_mutant",start.position[i]:end.position[i]]
}

store.seqs
```

Finally lets output all these sequences to a FASTA file for further analysis with the IEDB HLA binding prediction website.

```{r}
## Output a FASTA file for further analysis
write.fasta(seqs=store.seqs, ids=mutant.names, file="subsequences.fa")
```

# Patient HLA typing results and HLA binding prediction

> Q2: Identify 9-mer peptides in the identified sequence regions unique to the tumor that can be potentially presented to T cells.

Hint: Use the IEDB HLA binding prediction server above to identify the top ranked 9-mer peptides for each patient HLA (see above for HLA typing results).

> Q3: Identify the top peptide for each patient HLA allele (see above).

Hint: You can download a CSV formated result file for all predictions and use R or a spreadsheet application to identify the top ranked peptides for each allele. 

```{r}
HLA_predictions <- read.csv("result.csv")
head(HLA_predictions)
```

```{r}
# get the top 10 ranked peptides
HLA_top_ranks <- HLA_predictions[1:10, ]
View(HLA_top_ranks)
```

